src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    internal sealed class CurlResponseMessage : HttpResponseMessage
+    {
+        private readonly CurlResponseStream _responseStream = new CurlResponseStream();
+
+        #region Properties
+
+        internal CurlResponseStream ContentStream
+        {
+            get { return _responseStream; }
+        }
+
+        #endregion
+
+        internal CurlResponseMessage(HttpRequestMessage request)
+        {
+            RequestMessage = request;
+            Content = new StreamContent(_responseStream);
+        }
+    }
+
+        internal sealed class CurlResponseStream : Stream
+        {
+            private volatile bool _disposed = false;
+            private Stream _innerStream;
+            private readonly object _lockObject = new object();
+            private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
```

Comment:
To save a little bit of space and an allocation, instead of having _lockObject consider having:
```C#
private object LockObject { get { return _cancellationTokenSource; } }
```
and then using LockObject instead of _lockObject in the locking code.